File: bindings/wasm/examples/decode_staked_iota.mjs
// Copyright (c) 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// Decode `StakedIota` objects into typed JavaScript values.
//
// The GraphQL client returns each object's contents as raw BCS bytes. A single
// `StakedIota.tryFromObject(obj)` call gives typed, named-field access to
// id / poolId / stakeActivationEpoch / principal.

import {
  Direction,
  GraphQlClient,
  ObjectFilter,
  PaginationFilter,
  StakedIota,
  TransactionsFilter,
  initAsync,
} from "@iota/sdk-wasm";

await initAsync();

const client = GraphQlClient.newTestnet();

// Filtering objects by type alone scans every object on the network, which the
// GraphQL server rejects with a timeout. Pick a recent staker and filter by
// owner as well, so only that address' objects are looked at.
const stakers = await client.transactions(
  TransactionsFilter.new({ function: "0x3::iota_system::request_add_stake" }),
  PaginationFilter.new({ direction: Direction.Backward, limit: 1 }),
);
const staker = stakers.data.at(-1)?.transaction.sender();

if (staker === undefined) {
  console.log("No staking transactions on testnet right now.");
} else {
  console.log(`Latest staker: ${staker.toHex()}\n`);

  const page = await client.objects(
    ObjectFilter.new({
      typeTag: "0x3::staking_pool::StakedIota",
      owner: staker,
    }),
  );

  if (page.data.length === 0) {
    console.log(`No StakedIota objects owned by ${staker.toHex()} right now.`);
  } else {
    console.log(`Decoded ${page.data.length} StakedIota object(s):\n`);
    let totalPrincipal = 0n;
    for (const obj of page.data) {
      const staked = StakedIota.tryFromObject(obj);
      totalPrincipal += staked.principal();
      console.log(`- id:               ${staked.id().toHex()}`);
      console.log(`  pool_id:          ${staked.poolId().toHex()}`);
      console.log(`  stake_activation_epoch: ${staked.stakeActivationEpoch()}`);
      console.log(`  principal (nanos): ${staked.principal()}`);
      console.log();
    }

    console.log(`Total principal across page: ${totalPrincipal} nanos`);
  }
}

File: bindings/python/examples/decode_staked_iota.py
# Copyright (c) 2026 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

# Decode `StakedIota` objects into typed Python values.
#
# The GraphQL client returns each object's contents as raw BCS bytes. A single
# `StakedIota.try_from_object(obj)` call gives typed, named-field access to
# id / pool_id / stake_activation_epoch / principal.

from lib.iota_sdk import *

import asyncio


async def main():
    client = GraphQlClient.new_testnet()

    owner = Address.from_hex(
        "0xda1820edf693ee32b5729907b9b2ec8e64980ee8c008c17e89cfb4e5ecd72151")

    page = await client.objects(filter=ObjectFilter(
        type_tag="0x3::staking_pool::StakedIota", owner=owner))

    if len(page.data) == 0:
        print(f"No StakedIota objects owned by {owner.to_hex()} right now.")
        return

    print(f"Decoded {len(page.data)} StakedIota object(s):\n")
    total_principal = 0
    for obj in page.data:
        staked = StakedIota.try_from_object(obj)
        total_principal += staked.principal()
        print(f"- id:               {staked.id().to_hex()}")
        print(f"  pool_id:          {staked.pool_id().to_hex()}")
        print(f"  stake_activation_epoch: {staked.stake_activation_epoch()}")
        print(f"  principal (nanos): {staked.principal()}")
        print()

    print(f"Total principal across page: {total_principal} nanos")


if __name__ == "__main__":
    asyncio.run(main())
